src/pybind/mgr/dashboard/controllers/host.py

```diff
@@ -1,12 +1,79 @@
 from .. import mgr
-from ..security import Scope
+from ..exceptions import DashboardException
+from ..security import Permission, Scope
+from ..services.orchestrator import OrchClient
+from ..services.exception import handle_orchestrator_error
+
+
+def host_task(name, metadata, wait_for=10.0):
+    return Task("host/{}".format(name), metadata, wait_for)
+
+
+def merge_with_orchestrator(mgr_hosts):
+    """Merge mgr hosts with orchestrator hosts.
+
+    :param mgr_hosts: hosts returned from mgr
+    :return list of dict with host information
+    """
+    all_hosts = copy.copy(mgr_hosts)
+    orch = OrchClient()
+    if orch.available():
+        mgr_hostnames = {host['hostname'] for host in mgr_hosts}
+        orch_hosts = (dict(hostname=host.name, ceph_version='', services=[])
+                      for host in orch.hosts.list()
+                      if host.name not in mgr_hostnames)
+        all_hosts.extend(orch_hosts)
+    return all_hosts
 
 
 @ApiController('/host', Scope.HOSTS)
 class Host(RESTController):
+
     def list(self):
-        return mgr.list_servers()
+        hosts = mgr.list_servers()
+        if self._has_permissions(Permission.READ, Scope.ORCHESTRATOR):
+            return merge_with_orchestrator(hosts)
+        return hosts
+
+    @host_task('add', {'hostname': '{hostname}'})
+    @handle_orchestrator_error('host')
+    def create(self, hostname):
+        orch_client = OrchClient()
+        self._check_orchestrator_host_op(orch_client, hostname, True)
+        orch_client.hosts.add(hostname)
+
+    @host_task('remove', {'hostname': '{hostname}'})
+    @handle_orchestrator_error('host')
+    def delete(self, hostname):
+        orch_client = OrchClient()
+        self._check_orchestrator_host_op(orch_client, hostname, False)
+        orch_client.hosts.remove(hostname)
+
+    def _check_orchestrator_host_op(self, orch_client, hostname, add_host=True):
+        """Check if we can adding or removing a host with orchestrator
+
+        :param orch_client: Orchestrator client
+        :param add: True for adding host operation, False for removing host
+        :raise DashboardException
+        """
+        if not orch_client.available():
+            raise DashboardException(code='orchestrator_status_unavailable',
+                                     msg='Orchestrator is unavailable',
+                                     component='orchestrator')
+        host = orch_client.hosts.get(hostname)
+        if add_host and host:
+            raise DashboardException(
+                code='orchestrator_add_existed_host',
+                msg='{} is already in orchestrator'.format(hostname),
+                component='orchestrator')
+        if not add_host and not host:
+            raise DashboardException(
+                code='orchestrator_remove_nonexistent_host',
+                msg='Remove a non-existent host {} from orchestrator'.format(
+                    hostname),
+                component='orchestrator')
```

Comment:
Hm, I'd think the orchestrator cli would also benefit from these checks. Might it make sense to move them into the orch?

Reply:
This is checked both in frontend and backend. Yes, it will be good if orchestrator can check this too. I should be done with another PR.